pilot/pkg/config/mcp/controller.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package coredatamodel
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"istio.io/istio/pilot/pkg/model"
+	mcpclient "istio.io/istio/pkg/mcp/client"
+)
+
+var (
+	errNotFound      = errors.New("item not found")
+	errAlreadyExists = errors.New("item already exists")
+	errUnsupported   = errors.New("this operation is not supported by mcp controller")
+)
+
+//go:generate counterfeiter -o fakes/logger.go --fake-name Logger . logger
+type logger interface {
+	Warnf(template string, args ...interface{})
+	Infof(template string, args ...interface{})
+}
+
+type CoreDataModel interface {
+	model.ConfigStoreCache
+	mcpclient.Updater
+}
+
+type changeEvent struct {
+	metaName       string
+	resource       proto.Message
+	descriptorType string
+}
+
+type Controller struct {
+	configStore              map[string]map[string]*sync.Map
+	eventCh                  chan func(model.Config, model.Event)
+	updateCh                 chan changeEvent
+	descriptorsByMessageName map[string]model.ProtoSchema
+	logger                   logger
+}
+
+func NewController(logger logger) CoreDataModel {
+	descriptorsByMessageName := make(map[string]model.ProtoSchema, len(model.IstioConfigTypes))
+	for _, config := range model.IstioConfigTypes {
+		descriptorsByMessageName[config.MessageName] = config
+	}
+
+	configStore := make(map[string]map[string]*sync.Map)
+	for _, typ := range model.IstioConfigTypes.Types() {
+		configStore[typ] = make(map[string]*sync.Map)
+	}
+	return &Controller{
+		configStore:              configStore,
+		eventCh:                  make(chan func(model.Config, model.Event)),
+		updateCh:                 make(chan changeEvent),
+		descriptorsByMessageName: descriptorsByMessageName,
+		logger: logger,
+	}
+}
+
+func (c *Controller) Run(stop <-chan struct{}) {
```

Comment:
So how many goroutines are involved here ? There is grpc one that is reading the messages from the stream. Do we need another run ? Can it just make a callback when an object is received, with the proto it got from the stream ?

Reply:
This run will always run as a single goroutine regardless of the number of MCP clients. I don't think there is a need for another run, proto stream is handled through the `updateCh`.

Reply:
Agreed, I will update to reflect the changes we discussed  above :)